design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,269 @@
+pattern in our managed resources gets more important every day. This design document proposes some guiding principles 
+for designing managed resources as well as a few examples that show how these can map to actual resources.
+
+A few questions that this document needs to help answering:
+- What goes into `Spec` and what goes into `Status` of a managed resource?
+- Should _all_ information from the provider be represented in the Custom Resource?
+- What parts of the Custom Resource should be allowed to be consumed by the consumers?
+- Do we represent _desired_ state, _observed_ state or both?
+- How should we differentiate the fields that are used by provider from Crossplane's fields?
+
+This document tries to capture how the bridge between Crossplane and provider is shaped. However, each provider can
+wildly vary between each other. So, treat this document as a starting point, try to stick to it for consistency but
+deviate if required and feel welcome to come back and improve with your observations.
+
+## Guiding Principles
+
+Our users are the stakeholders on how we make decisions. There are two types of users for managed resources:
+* Input Users: Users who configure/create/manage the resource. Crossplane Claim Controller, Cluster Admin are a few
+  examples of this type. We want our `Spec` to be sole place for them to configure everything about the resource.
+* Output Users: Users who refer to/query/use/extract information from the managed resource. We want them to be able to
+  perform all read-only operations on the given resource as if they have the cloud provider's API available. This can
+  include configurations and metadata information about the resource. We also want them to access only `Status`
+  sub-resource of the managed resource because this is where the most up-to-date information about the resource is
+  stored and we are able to limit RBAC of these users to access only `Status`.
+
+### Naming Fields
+
+The thinking process of naming decision could respect provider's decision as the starting point and deviate from that
+if there is a very good reason to. Naming parity makes inter-resource references easier and lowers the entry barrier for
+both developers and users. This is especially handy in cases where Crossplane does not yet support the _referred_ resource
+or user may not want some of their resources in their Crossplane environment.
+
+Ideally, for the resources that do have Crossplane representations, we'll have `ObjectReference` type for
+these fields. In that case, you can append `Ref` to the field name, i.e. `network (string)` -> `networkRef (ObjectReference)`
+
+In some cases, we might have collisions or cases where provider field name is too similar to another field in the
+managed resource's Custom Resource fields. This is partly solved in [Consumer-based Struct Levels](#consumer-based-struct-levels)
+section.
+
+### High Fidelity
```

Comment:
This part tries to tackle that:
```
For example, `network` could be a top level information in `GET` request of a database resource but the details of that
network should not be included via making extra calls, we want only the information that is present on the top level.
If there is a reference to another provider API object (the object could be a tree of other objects like in Azure), only
include the provider's identifier of that object to make it easier to query it when needed. Trying to include the whole
tree can even go to the state where all your setup that is connected to each other is represented under one resource's
status, which is a very bad UX.
```

Reply:
Yes, @negz you are correct. There is not currently an issue open specifically for this Azure case. One thing that could be difficult with your suggestion here @muvaf is that it is not always readily apparent which information is "top level" for the apis, and it can require some digging and testing to determine if a nested object can actually be applied at creation time.